src/engine.ts

```diff
@@ -0,0 +1,207 @@
+  [ElementKind[ElementKind.heading5]]: 4,
+  [ElementKind[ElementKind.heading6]]: 4,
+  [ElementKind[ElementKind.paragraph]]: 3,
+  [ElementKind[ElementKind.preformatted]]: 5,
+  [ElementKind[ElementKind.strong]]: 6,
+  [ElementKind[ElementKind.em]]: 6,
+  [ElementKind[ElementKind["list-item"]]]: 1,
+  [ElementKind[ElementKind["o-list-item"]]]: 1,
+  [ElementKind[ElementKind["group-list-item"]]]: 1,
+  [ElementKind[ElementKind["group-o-list-item"]]]: 1,
+  [ElementKind[ElementKind.image]]: 1,
+  [ElementKind[ElementKind.embed]]: 1,
+  [ElementKind[ElementKind.hyperlink]]: 3,
+  [ElementKind[ElementKind.label]]: 4,
+  [ElementKind[ElementKind.span]]: 7,
+};
+
+export interface Tree {
+  key: string;
+  children: Node[];
+}
+
+export interface Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+}
+
+class TextNode implements Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+
+  constructor(start: number, end: number, text: string) {
+    this.key = uuid();
+    this.start = start;
+    this.end = end;
+    this.tag = ElementKind[ElementKind.span];
+    this.text = text;
+    this.children = [];
+  }
+}
+
+export interface RichTextSpan {
+  start: number;
+  end: number;
+  type: string;
+  text: string;
+}
+
+export interface RichTextBlock {
+  text: string;
+  spans: RichTextSpan[];
+}
+
+export interface Group {
+  elected: Node;
+  others: Node[];
+}
+
+interface SlicedNode {
+  inner: Node;
+  outer?: Node;
+}
+
+function sliceNode(text: string, elected: Node, node: Node): SlicedNode {
+  if (node.start < elected.start) {
+    return {
+      inner: R.merge(node, {
+        key: uuid(),
+        start: elected.start,
+        text: text.slice(elected.start, node.end),
+      }),
+      outer: R.merge(node, {
+        key: uuid(),
+        end : elected.start,
+        text: text.slice(node.start, elected.start),
+      }),
+    };
+  } else if (node.end > elected.end) {
+      return {
+        inner: R.merge(node, {
+          key: uuid(),
+          end: elected.end,
+          text: text.slice(node.start, elected.end),
+        }),
+        outer: R.merge(node, {
+          key: uuid(),
+          start: elected.end,
+          text: text.slice(elected.end, node.end),
+        }),
+      };
+  } else {
+    return {
+      inner: node,
+    };
+  }
+}
+
+function partitionGroup(text: string, group: Group): Node[] {
+  const partitioned = group.others.reduce(({ inner: innerAcc, outer: outerAcc }, node) => {
+    const slicedNode = sliceNode(text, group.elected, node);
+    return {
+      inner: innerAcc.concat(slicedNode.inner),
+      outer: slicedNode.outer ? outerAcc.concat(slicedNode.outer) : outerAcc,
+    };
+  }, { inner: new Array<Node>(), outer: new Array<Node>()});
+
+  const { inner, outer } = partitioned;
+  const head = R.assoc('children', buildTree(text, inner), group.elected);
```

Comment:
It seems to be close to the old algo in there. I think that the inner array will have the same problem, you should reprocess them to split nodes on the second level etc.